### ISSUE TITLE
Stack allocated optimization 

### ISSUE THREAD

**Issue by aaowens** [JuliaNLSolvers/Optim.jl#459]
ForwardDiff now supports computing derivatives as StaticArrays and storing the results in an ImmutableDiffResult (https://github.com/JuliaDiff/ForwardDiff.jl/pull/245). For cheap low-dimensional functions this is quite a bit faster than it is with Arrays. I have an example here of a non-allocating BFGS method using these features (requires ForwardDiff master).
```
using Optim, ForwardDiff, StaticArrays, BenchmarkTools
function my_optBFGS(f, x::AbstractVector)
  res = DiffBase.GradientResult(x)
  res = ForwardDiff.gradient!(res, f, x)
  fx = DiffBase.value(res)
  isfinite(fx) || return (NaN*x, NaN)
  jx = DiffBase.gradient(res)
  hx = diagm(ones(x))
  tol = 1e-8
  for n = 1:100
    alpha = 1.
    norm(jx) < tol && return (x, fx)
    d = hx\jx
    propose = x - alpha*d
    fpropose = f(propose)
    # implement back-tracking line-search with Armijo–Goldstein condition
    # Reduce step size until decrease is acceptable relative to expected decrease
    dotdj = dot(d, jx)
    countme = 1
    while countme < 100 &&(!isfinite(fpropose) || fpropose > fx - 0.5*alpha*dotdj)
      alpha = 0.8*alpha
      propose = x - alpha*d
      fpropose = f(propose)
      countme += 1
    end
    s = -alpha*d
    res = ForwardDiff.gradient!(res, f, propose)
    j2 = DiffBase.gradient(res)
    y = j2 - jx
    hx = hx + y*y' / (y'*s) - (hx*(s*s')*hx)/(s'*hx*s)
    x = propose
    jx = j2; fx = DiffBase.value(res)
  end
  return (NaN*x, NaN)
end
rosenbrock(x) =  (1.0 - x[1])^2 + 100.0 * (x[2] - x[1]^2)^2
x = zeros(2); sx = @SVector zeros(2)
od = OnceDifferentiable(rosenbrock, x; autodiff = :forward);
ot = BFGS();
julia> result = @btime optimize($od, $x, $ot);
  85.005 μs (1359 allocations: 26.22 KiB)

julia> result = @btime optimize(rosenbrock, $x, $ot);
  90.036 μs (1256 allocations: 24.16 KiB)

julia> sresult = @btime my_optBFGS(rosenbrock, $sx);
  1.606 μs (0 allocations: 0 bytes)

julia>  (result.minimizer - sresult[1])
2-element SVector{2,Float64}:
 -7.39614e-9
 -1.47983e-8

```

The stack-allocated version is about 50x faster than Optim. I think this is a combination of not needing to allocate memory, SIMD optimized gradient computations in ForwardDiff, and the faster matrix operations from StaticArrays.

It would be really cool if Optim could do this automatically if the initial guess is a StaticArray. I'm not sure how hard it would be. The trick with the DiffResults is to re-alias the input in "mutating" statements like `res = ForwardDiff.gradient!(res, f, x)` , so it works whether `res` is mutable or immutable. 

If it's not practical for every algorithm to implement this, maybe there could be `StaticNewton()` or `StaticBFGS()` algorithms just for low-dimensional problems.

**Comment by ChrisRackauckas** [comment]
BTW I have experience doing this because DifferentialEquations.jl has inplace and out-of-place versions for this purpose. It does require a pretty significant re-write of the algorithms, usually a completely separate version of the inner loop if everything was written to use inplace operations and caching appropriately. But yes, it would be nice to see this, but it isn't a small amount of code to support it.

**Comment by pkofod** [comment]
Keep in mind that when you find that your code is 50x faster than Optim you have to be a bit careful. Is the line search step comparable, that is do you have the same number of objective and gradient evaluations as we do? Also keep in mind that as a general purpose package, we have various extra steps that might add time, especially for such a cheap function.

Anyway, I think the idea is interesting, but I personally don't have the time right now to prioritize it. However, let's discuss it ! It might be useful to some.

**Comment by aaowens** [comment]
My method had 71 objective calls and 22 gradient calls, while Optim used 53 of both. This means Optim is doing somewhat more work, but not enough to explain the gap.

I often need to optimize small functions millions of times for different parameters, which is why I started using this code instead of Optim. This happens in economics research where you might need to solve a simple optimization problem for each point in a large state space.

**Comment by anriseth** [comment]
I am also doing thousands of calls to optimize with low-dimensional problems quite often, so this is interesting. As a side note: If you stay with your code, you will probably benefit from implementing an interpolating backtracking algorithm instead of vanilla (see `LineSearches.jl`s `BackTracking` for a simple implementation of order 2 and order 3 interpolation)

> My method had 71 objective calls and 22 gradient calls, while Optim used 53 of both. This means Optim is doing somewhat more work, but not enough to explain the gap.

You're comparing Optim BFGS with Hagerzhang linesearch, to your BFGS with vanilla backtracking. And your tolerances are different (inf-norm vs l2-norm).

**Comment by pkofod** [comment]
For what it's worth, `BFGS` with the quadratic backtracking in LineSearches gives me
```julia
julia> result
Results of Optimization Algorithm
 * Algorithm: BFGS
 * Starting Point: [0.0,0.0]
 * Minimizer: [0.9999999999992523,0.9999999999982809]
 * Minimum: 5.558756e-24
 * Iterations: 23
 * Convergence: true
   * |x - x'| < 1.0e-32: false 
     |x - x'| = 1.13e-09 
   * |f(x) - f(x')| / |f(x)| < 1.0e-32: false
     |f(x) - f(x')| / |f(x)| = NaN 
   * |g(x)| < 1.0e-08: true 
     |g(x)| = 8.79e-11 
   * stopped by an increasing objective: false
   * Reached Maximum Number of Iterations: false
 * Objective Calls: 31
 * Gradient Calls: 24
```
that is two additional gradient calls, but half the objective calls.

**Comment by ChrisRackauckas** [comment]
>I am also doing thousands of calls to optimize with low-dimensional problems quite often, so this is interesting. 

Well, a form of pre-allocating the caches is really what's needed then.

**Comment by pkofod** [comment]
I believe we are almost there for many (most?) algorithms as long as you provide the cache as the last positional argument to optimize.

**Comment by aaowens** [comment]
> Well, a form of pre-allocating the caches is really what's needed then. In theory there should be a form where Optim.jl doesn't need to allocate anything at all since optimization is an iterative updating algorithm which can re-use the same cache and doesn't need to save history.

I don't think lack of pre-allocating is the problem here. Quoting from https://github.com/JuliaDiff/ForwardDiff.jl/pull/215 . we saw

```
# non-allocating old method
julia> @btime ForwardDiff.gradient!($out, sum, $x, $cfg);
  666.856 ns (0 allocations: 0 bytes)

# stack-allocated new method
julia> @btime ForwardDiff.gradient(sum, $sx);
  9.394 ns (0 allocations: 0 bytes)
```
Both have zero allocations, but the stack-allocated method is much faster.

**Comment by pkofod** [comment]
> I don't think lack of pre-allocating is the problem here.

I think you can find cases where it is and where it isn't. Though, I am very much motivated to do this, I'm just prioritizing cleaning up NLsolve a bit these days, as there are things in there I **do not want** when Julia v1.0 hits the streets. This is more of a nice to have, but I *do* agree that it is nice indeed.

**Comment by aaowens** [comment]
I used the cubic backtracking linesearch code from LineSearches.jl to upgrade my code and put it in a package. It's available here: https://github.com/aaowens/StaticOptim.jl

It's a BFGS solver which doesn't allocate. I think it could be useful for people solving a lot of these tiny optimization problems.

**Comment by anriseth** [comment]
@aaowens good to see that things work:) Do you think it will be easy to update the BackTracking code in `LineSearches.jl` with your changes? If so, we'd appreciate a PR ;)

**Comment by aaowens** [comment]
I don't think my code is generic enough for `LineSearches.jl`. I replaced all the generic things like references to `NLSolversBase` with the simplest alternative that worked with a `StaticArray`. I also wasn't able to get the code to not allocate when I used a linesearch function. It stopped allocating when I placed the whole linesearch function body inside the optimization routine (`@inline` didn't seem to help). This included swapping out references to `make_ϕ` and `ϕ(α)` with `f(x + α*s)`.
